feat: add `stats/base/dists/bradford/skewness`

vivekmaurya001 opened the pull request:
Progresses #168.

This pull request:

-    add `stats/base/dists/bradford/skewness`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #{{TODO: add issue number}} 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

 - [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)


* * *

@stdlib-js/reviewers 

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

vivekmaurya001 committed:
feat: add stats/base/dists/bradford/skewness

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: passed
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

vivekmaurya001 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

vivekmaurya001 commented:
currently there are 10 test cases which fail at tolerance factor of 50, but these cases persists even I increase the factor to 1e4


![Screenshot from 2025-03-26 21-09-18](https://github.com/user-attachments/assets/849c6204-9e64-4144-9138-49c30545a7cb)

vivekmaurya001 commented:
@anandkaranubc , @kgryte , Any suggestion on this

anandkaranubc commented:
@vivekmaurya001, that's the main reason I didn’t submit my PRs on skewness and kurtosis — they lead to large divergent results due to ("My Hypothesis"):

1. A lot of calculations being performed, which causes floating-point precision errors.  
2. The shape parameter being very small, leading to differences in the `ln` calculations compared to the reference (SciPy) implementation.  
3. The expected values being very small.

These are my local results:

- For small `c`: `tol = 930226.0 * EPS * abs(expected[i])`
- For large `c`: `tol = 600.0 * EPS * abs(expected[i])`

There are actually two ways to implement this:

```js
k = ln(1.0 + c);
out = SQRT2;
out *= ( (12.0*c*c) - (9.0*k*c*(c+2.0)) + (2.0*k*k*(c*(c+3.0)+3.0)) );
out /= sqrt(c * (c*(k-2.0) + (2.0*k)));
out /= ( (3.0*c*(k-2.0)) + (6.0*k) );
return out;
```

or

```js
var g1 = SQRT2 * (12*c*c - 9*c*k*(c+2) + 2*k*k*(c*(c+3)+3));
g1 /= sqrt(c*(c*(k-2)+2*k)) * (3*c*(k-2) + 6*k);
return g1;
```

The second one gives better results than the first. Here, better results mean less divergence from SciPy. 

That said, I think we might need to check if there have been any changes in the upstream implementations of `ln` or `log1p`, and update ours if needed.

cc: @kgryte

vivekmaurya001 commented:
Yess, Also the scipy docs are using log(1+c) instead of log1p

anandkaranubc commented:
> Yess, Also the scipy docs are using log(1+c) instead of log1p

Would you mind checking the upstream implementation of `ln` to see if there's any difference? If there is, create an issue and we can sync up on that.

[FreeBSD implementation ](https://svnweb.freebsd.org/base/release/12.2.0/lib/msun/src/e_log.c?view=markup&sortby=date)
[Our implementation ](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/ln/lib/main.js)

vivekmaurya001 commented:
Sure I will try it first

vivekmaurya001 commented:
@anandkaranubc , I found some slight differences in lg2, lg3 etc values in free bsd and our, I tried correcting it and running but still not improved, apart from this all looked same

anandkaranubc commented:
> @anandkaranubc , I found some slight differences in lg2, lg3 etc values in free bsd and our, I tried correcting it and running but still not improved, apart from this all looked same

~~Could you create a separate issue for this so we can tackle it there and get some feedback?~~ It might also be worthwhile to check how SciPy has implemented `ln`.

Edit: Are you sure about that?

The values seem the same to me:

https://github.com/stdlib-js/stdlib/blob/14a166db9d76fe6663adcdd695769cacc438dadc/lib/node_modules/%40stdlib/math/base/special/ln/scripts/evalpoly.js#L40-L51

vivekmaurya001 commented:
Oops yes, I realised some last digits were not matching

vivekmaurya001 committed:
feat: add stats/base/dists/bradford/skewness

vivekmaurya001 committed:
Merge branch 'bradford/skewness' of https://github.com/vivekmaurya001/stdlib into bradford/skewness

vivekmaurya001 committed:
added reason for test failures

vivekmaurya001 commented:
Added `ln` logic instead of `log1p` and TODO comment in test file

Planeshifter commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into bradford/skewness

Planeshifter committed:
chore: minor clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
chore: minor clean-up and add comment

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter reviewed (approved):
Thanks, @vivekmaurya001!

We needed large tolerances for quite a few of the Bradford packages due to numerical differences accumulating from the various operations. I updated the tolerance so that tests pass and will merge once CI has cleared.
